File: blue-core/src/blue/noteProcessor/LineMultiplyProcessor.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2000-2003 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */

package blue.noteProcessor;

import blue.BlueSystem;
import blue.plugin.NoteProcessorPlugin;
import blue.soundObject.Note;
import blue.soundObject.NoteList;
import blue.soundObject.NoteParseException;
import electric.xml.Element;
import java.io.Serializable;

/**
 * @author steven
 * 
 */
@NoteProcessorPlugin(displayName="LineMultiplyProcessor", position = 150)
public class LineMultiplyProcessor implements NoteProcessor, Serializable {

    private String lineMultiplyString = "0 0";

    private int pfield = 4;

    /*
     * (non-Javadoc)
     * 
     * @see blue.noteProcessor.NoteProcessor#processNotes(blue.soundObject.NoteList)
     */
    @Override
    public void processNotes(NoteList in) throws NoteProcessorException {
        Note temp;
        float oldVal = 0f;
        float multiplyVal = 0f;
        ValueTimeMapper tm = ValueTimeMapper
                .createValueTimeMapper(this.lineMultiplyString);

        if (tm == null) {
            throw new NoteProcessorException(this, BlueSystem
                    .getString("noteProcessorException.lineMultStringErr"),
                    pfield);
        }

        for (int i = 0; i < in.size(); i++) {
            temp = in.get(i);
            try {
                oldVal = Float.parseFloat(temp.getPField(this.pfield));
                multiplyVal = tm.getValueForBeat(temp.getStartTime());
            } catch (NumberFormatException ex) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.pfieldNotFloat"),
                        pfield);
            } catch (Exception ex) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.missingPfield"),
                        pfield);
            }

            if (Float.isNaN(multiplyVal)) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.noteBeatErr"),
                        pfield);
            }
            temp.setPField(Float.toString(oldVal * multiplyVal), this.pfield);

        }
    }

    @Override
    public String toString() {
        return "[line multiply]";
    }

    public static void main(String[] args) {
        NoteList n = new NoteList();

        for (int i = 0; i < 10; i++) {
            try {
                n.add(Note.createNote("i1 " + i + " 1 3 4"));
            } catch (NoteParseException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

        System.out.println("before: \n\n" + n + "\n\n");

        LineMultiplyProcessor lmp = new LineMultiplyProcessor();
        lmp.setLineMultiplyString("0 0 5 1 9 0");
        try {
            lmp.processNotes(n);
        } catch (NoteProcessorException ex) {
            System.out.println("Exception: " + ex.getMessage());
        }

        System.out.println("after: \n\n" + n + "\n\n");
    }

    public String getPfield() {
        return Integer.toString(pfield);
    }

    public void setPfield(String pfield) {
        this.pfield = Integer.parseInt(pfield);
    }

    public String getLineMultiplyString() {
        return lineMultiplyString;
    }

    public void setLineMultiplyString(String string) {
        lineMultiplyString = string;
    }

    public static NoteProcessor loadFromXML(Element data) {
        LineMultiplyProcessor lmp = new LineMultiplyProcessor();

        lmp.setPfield(data.getElement("pfield").getTextString());
        lmp.setLineMultiplyString(data.getElement("lineMultiplyString")
                .getTextString());

        return lmp;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.noteProcessor.NoteProcessor#saveAsXML()
     */
    @Override
    public Element saveAsXML() {
        Element retVal = new Element("noteProcessor");
        retVal.setAttribute("type", this.getClass().getName());

        retVal.addElement("pfield").setText(this.getPfield());
        retVal.addElement("lineMultiplyString").setText(
                this.getLineMultiplyString());

        return retVal;
    }

}

File: packages/blue-data/src/note-processors/line-multiply-processor.ts
import { NoteProcessor } from './note-processor';
import { NoteProcessorException } from './note-processor-exception';
import { NoteList } from '../sound-objects/note-list';
import { Element } from '../serialization/xml-reader';
import { ValueTimeMapper } from './value-time-mapper';

const JAVA_TYPE = 'blue.noteProcessor.LineMultiplyProcessor';

export class LineMultiplyProcessor extends NoteProcessor {
  private _lineMultiplyString = '0 1';
  private _pfield = 4;

  constructor();
  constructor(src: LineMultiplyProcessor);
  constructor(src?: LineMultiplyProcessor) {
    super();
    if (src) {
      this._lineMultiplyString = src._lineMultiplyString;
      this._pfield = src._pfield;
    }
  }

  getPfield(): string {
    return this._pfield.toString();
  }
  setPfield(pfield: string): void {
    this._pfield = parseInt(pfield, 10);
  }

  getLineMultiplyString(): string {
    return this._lineMultiplyString;
  }
  setLineMultiplyString(lineMultiplyString: string): void {
    this._lineMultiplyString = lineMultiplyString;
  }

  override process(notes: NoteList): NoteList {
    const tm = ValueTimeMapper.createValueTimeMapper(this._lineMultiplyString);
    if (tm === null) {
      throw new NoteProcessorException('Error in line multiply string', this._pfield);
    }

    for (const note of notes) {
      let oldVal: number;
      let multiplyVal: number;
      try {
        oldVal = parseFloat(note.getPField(this._pfield)!);
        multiplyVal = tm.getValueForBeat(note.getStartTime());
      } catch {
        throw new NoteProcessorException('Pfield is not a double', this._pfield);
      }
      if (isNaN(oldVal)) {
        throw new NoteProcessorException('Pfield is not a double', this._pfield);
      }
      if (isNaN(multiplyVal)) {
        throw new NoteProcessorException('Note beat out of range', this._pfield);
      }
      note.setPField((oldVal * multiplyVal).toString(), this._pfield);
    }
    return notes;
  }

  override getDisplayName(): string {
    return 'LineMultiplyProcessor';
  }

  override deepCopy(): LineMultiplyProcessor {
    return new LineMultiplyProcessor(this);
  }

  saveAsXML(): Element {
    const elem = new Element('noteProcessor');
    elem.setAttribute('type', JAVA_TYPE);
    elem.addElement('pfield').setText(this.getPfield());
    elem.addElement('lineMultiplyString').setText(this.getLineMultiplyString());
    return elem;
  }

  static loadFromXML(data: Element): LineMultiplyProcessor {
    const proc = new LineMultiplyProcessor();
    const pf = data.getTextString('pfield');
    if (pf !== null) proc._pfield = parseInt(pf, 10);
    const lms = data.getTextString('lineMultiplyString');
    if (lms !== null) proc._lineMultiplyString = lms;
    return proc;
  }
}
